com.hazelcast.internal.crdt.pncounter.PNCounterConsistencyLostTest.consistencyLostExceptionIsThrownWhenTargetReplicaDisappears

_master_ (commit 4f848dc1a2f740dcda30c7f9ce36aec5ed9ce52e)

Failed on OpenJDK: https://jenkins.hazelcast.com/job/Hazelcast-master-OpenJDK-sts/17/testReport/junit/com.hazelcast.internal.crdt.pncounter/PNCounterConsistencyLostTest/consistencyLostExceptionIsThrownWhenTargetReplicaDisappears/
<details><summary>Stacktrace:</summary>

```
java.lang.Exception: Unexpected exception, expected<com.hazelcast.core.ConsistencyLostException> but was<java.lang.IllegalStateException>
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:30)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.IllegalStateException: Attempt to reuse the same operation in multiple invocations. Operation is com.hazelcast.internal.crdt.pncounter.operations.AddOperation{serviceName='hz:impl:PNCounterService', identityHash=1136223929, partitionId=-1, replicaIndex=0, callId=-4, invocationTime=1671124688343 (2022-12-15 17:18:08.343), waitTimeout=-1, callTimeout=60000, tenantControl=com.hazelcast.spi.impl.tenantcontrol.NoopTenantControl@0, name=counter}
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:545)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:241)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:71)
	at com.hazelcast.internal.crdt.pncounter.PNCounterProxy.invokeInternal(PNCounterProxy.java:195)
	at com.hazelcast.internal.crdt.pncounter.PNCounterProxy.invokeInternal(PNCounterProxy.java:205)
	at com.hazelcast.internal.crdt.pncounter.PNCounterProxy.invoke(PNCounterProxy.java:159)
	at com.hazelcast.internal.crdt.pncounter.PNCounterProxy.addAndGet(PNCounterProxy.java:100)
	at com.hazelcast.internal.crdt.pncounter.PNCounterConsistencyLostTest.mutate(PNCounterConsistencyLostTest.java:80)
	at com.hazelcast.internal.crdt.pncounter.AbstractPNCounterConsistencyLostTest.consistencyLostExceptionIsThrownWhenTargetReplicaDisappears(AbstractPNCounterConsistencyLostTest.java:44)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:19)
	... 4 more

```
</details>
<details><summary>Standard output:</summary>

```
Finished Running Test: driverCanContinueSessionByCallingReset in 0.449 seconds.
17:18:08,566  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [LifecycleService] Thread-42559 - [127.0.0.1]:5703 [dev] [5.3.0-SNAPSHOT] [127.0.0.1]:5703 is SHUTTING_DOWN
17:18:08,566  WARN |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [Node] Thread-42559 - [127.0.0.1]:5703 [dev] [5.3.0-SNAPSHOT] Terminating forcefully...
17:18:08,566  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [Node] Thread-42559 - [127.0.0.1]:5703 [dev] [5.3.0-SNAPSHOT] Shutting down connection manager...
17:18:08,566  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [MockServer] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5703, uuid=28862ca1-f11e-4910-bb80-75ff45b14919], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5702, uuid=2f268ffd-c35a-49d0-9be7-af98830b3f5d], remoteEndpoint=[address=[127.0.0.1]:5703, uuid=28862ca1-f11e-4910-bb80-75ff45b14919], alive=false}
17:18:08,566  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [MockServer] Thread-42559 - [127.0.0.1]:5703 [dev] [5.3.0-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5702, uuid=2f268ffd-c35a-49d0-9be7-af98830b3f5d], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5703, uuid=28862ca1-f11e-4910-bb80-75ff45b14919], remoteEndpoint=[address=[127.0.0.1]:5702, uuid=2f268ffd-c35a-49d0-9be7-af98830b3f5d], alive=false}
17:18:08,566  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [MembershipManager] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Removing Member [127.0.0.1]:5703 - 28862ca1-f11e-4910-bb80-75ff45b14919 lite
17:18:08,566  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [ClusterService] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] 

Members {size:1, ver:5} [
	Member [127.0.0.1]:5702 - 2f268ffd-c35a-49d0-9be7-af98830b3f5d this
]

17:18:08,566  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [Node] Thread-42559 - [127.0.0.1]:5703 [dev] [5.3.0-SNAPSHOT] Shutting down node engine...
17:18:08,566  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [TransactionManagerService] hz.PNCounterConsistencyLostTest_recursing_clifford.cached.thread-4 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5703, UUID: 28862ca1-f11e-4910-bb80-75ff45b14919
17:18:08,567  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [NodeExtension] Thread-42559 - [127.0.0.1]:5703 [dev] [5.3.0-SNAPSHOT] Destroying node NodeExtension.
17:18:08,568  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [Node] Thread-42559 - [127.0.0.1]:5703 [dev] [5.3.0-SNAPSHOT] Hazelcast Shutdown is completed in 2 ms.
17:18:08,568  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [LifecycleService] Thread-42559 - [127.0.0.1]:5703 [dev] [5.3.0-SNAPSHOT] [127.0.0.1]:5703 is SHUTDOWN
17:18:08,568  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [LifecycleService] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] [127.0.0.1]:5702 is SHUTTING_DOWN
17:18:08,568  WARN |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [Node] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Terminating forcefully...
17:18:08,568  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [Node] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Shutting down connection manager...
17:18:08,568  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [Node] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Shutting down node engine...
17:18:08,570  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [NodeExtension] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Destroying node NodeExtension.
17:18:08,570  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [Node] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Hazelcast Shutdown is completed in 2 ms.
17:18:08,570  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [LifecycleService] Thread-42559 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] [127.0.0.1]:5702 is SHUTDOWN
17:18:08,570  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [LifecycleService] Thread-42559 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
17:18:08,570  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [Node] Thread-42559 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Node is already shutting down... Waiting for shutdown process to complete...
17:18:08,570  INFO |consistencyLostExceptionIsThrownWhenTargetReplicaDisappears| - [LifecycleService] Thread-42559 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after consistencyLostExceptionIsThrownWhenTargetReplicaDisappears(com.hazelcast.internal.crdt.pncounter.PNCounterConsistencyLostTest): BuildInfo{version='5.3.0-SNAPSHOT', build='20221215', buildNumber=20221215, revision=4f848dc, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'consistencyLostExceptionIsThrownWhenTargetReplicaDisappears(com.hazelcast.internal.crdt.pncounter.PNCounterConsistencyLostTest):'
17:18:05, accumulated pauses: 236 ms, max pause: 191 ms, pauses over 1000 ms: 0


No metrics recorded during the test

```
</details>

--- a/hazelcast/src/main/java/com/hazelcast/internal/crdt/pncounter/PNCounterProxy.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/crdt/pncounter/PNCounterProxy.java
@@ -44,6 +44,7 @@ import java.util.List;
 import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
 
 import static com.hazelcast.internal.crdt.pncounter.PNCounterService.SERVICE_NAME;
+import static com.hazelcast.spi.impl.operationservice.OperationAccessor.cloneAndReset;
 
 /**
  * Member proxy implementation for a {@link PNCounter}.
@@ -199,9 +200,10 @@ public class PNCounterProxy extends AbstractDistributedObject<PNCounterService> 
         } catch (HazelcastException e) {
             logger.fine("Exception occurred while invoking operation on target " + target + ", choosing different target", e);
             if (excludedAddresses == EMPTY_ADDRESS_LIST) {
-                excludedAddresses = new ArrayList<Address>();
+                excludedAddresses = new ArrayList<>();
             }
             excludedAddresses.add(target);
+            operation = cloneAndReset(operation, getNodeEngine().getSerializationService());
             return invokeInternal(operation, excludedAddresses, e);
         }
     }

--- a/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/Operation.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/Operation.java
@@ -385,6 +385,11 @@ public abstract class Operation implements DataSerializable, Tenantable {
         onSetCallId(newId);
     }
 
+    // Accessed using OperationAccessor
+    final void resetCallId() {
+        CALL_ID.set(this, 0);
+    }
+
     /**
      * Called every time a new <code>callId</code> is set on the operation. A new
      * <code>callId</code> is set before initial invocation and before every

--- a/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/OperationAccessor.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/OperationAccessor.java
@@ -17,6 +17,7 @@
 package com.hazelcast.spi.impl.operationservice;
 
 import com.hazelcast.cluster.Address;
+import com.hazelcast.internal.serialization.SerializationService;
 import com.hazelcast.internal.server.ServerConnection;
 import com.hazelcast.spi.annotation.PrivateApi;
 
@@ -77,6 +78,21 @@ public final class OperationAccessor {
      */
     public static void setCallTimeout(Operation op, long callTimeout) {
         op.setCallTimeout(callTimeout);
+    }
+
+    /**
+     * Clone and reset the supplied operation.
+     *
+     * @param op the operation to clone and reset.
+     * @param serializationService the serialization service to use for cloning.
+     * @return a clone of the supplied operation, with its invocation time reset
+     * to -1 and its call ID reset to 0.
+     */
+    public static Operation cloneAndReset(Operation op, SerializationService serializationService) {
+        Operation clone = serializationService.toObject(serializationService.toData(op));
+        clone.setInvocationTime(-1);
+        clone.resetCallId();
+        return clone;
     }
 
 }

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/OperationAccessorTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/OperationAccessorTest.java
@@ -34,7 +34,9 @@ import static com.hazelcast.spi.impl.operationservice.OperationAccessor.setCallI
 import static com.hazelcast.spi.impl.operationservice.OperationAccessor.setCallTimeout;
 import static com.hazelcast.spi.impl.operationservice.OperationAccessor.setCallerAddress;
 import static com.hazelcast.spi.impl.operationservice.OperationAccessor.setInvocationTime;
-import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @RunWith(MockitoJUnitRunner.class)
 @Category({QuickTest.class, ParallelJVMTest.class})

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/OperationAccessorTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/OperationAccessorTest.java
@@ -1,0 +1,101 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl.operationservice;
+
+import com.hazelcast.cluster.Address;
+
+import com.hazelcast.internal.serialization.impl.DefaultSerializationServiceBuilder;
+import com.hazelcast.internal.server.ServerConnection;
+import com.hazelcast.spi.impl.operationservice.impl.DummyOperation;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import static com.hazelcast.spi.impl.operationservice.OperationAccessor.deactivate;
+import static com.hazelcast.spi.impl.operationservice.OperationAccessor.setCallId;
+import static com.hazelcast.spi.impl.operationservice.OperationAccessor.setCallTimeout;
+import static com.hazelcast.spi.impl.operationservice.OperationAccessor.setCallerAddress;
+import static com.hazelcast.spi.impl.operationservice.OperationAccessor.setInvocationTime;
+import static org.junit.jupiter.api.Assertions.*;
+
+@RunWith(MockitoJUnitRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class OperationAccessorTest {
+
+    @Mock
+    private ServerConnection serverConnection;
+    @Mock
+    private Address address;
+
+    @Test
+    public void testCloneAndResetOperation() {
+        Operation operation = new DummyOperation();
+
+        setCallId(operation, 1);
+        setInvocationTime(operation, 1000);
+        Operation clone = OperationAccessor.cloneAndReset(operation, new DefaultSerializationServiceBuilder().build());
+        assertEquals(0, clone.getCallId());
+        assertEquals(-1, clone.getInvocationTime());
+    }
+
+    @Test
+    public void testSetCallId() {
+        Operation operation = new DummyOperation();
+        setCallId(operation, 10);
+        assertEquals(10, operation.getCallId());
+    }
+
+    @Test
+    public void testSetInvocationTime() {
+        Operation operation = new DummyOperation();
+        setInvocationTime(operation, 10);
+        assertEquals(10, operation.getInvocationTime());
+    }
+
+    @Test
+    public void testSetCallerAddress() {
+        Operation operation = new DummyOperation();
+        setCallerAddress(operation, address);
+        assertEquals(address, operation.getCallerAddress());
+    }
+
+    @Test
+    public void testDeactivate() {
+        Operation operation = new DummyOperation();
+        setCallId(operation, 10);
+        assertTrue(deactivate(operation));
+        assertFalse(operation.isActive());
+    }
+
+    @Test
+    public void testSetCallTimeout() {
+        Operation operation = new DummyOperation();
+        setCallTimeout(operation, 10);
+        assertEquals(10, operation.getCallTimeout());
+    }
+
+    @Test
+    public void testSetConnection() {
+        Operation operation = new DummyOperation();
+        OperationAccessor.setConnection(operation, serverConnection);
+        assertEquals(serverConnection, operation.getConnection());
+    }
+}

